### ISSUE TITLE
URGENT! Ctx.Web.ApplyProvisioningTemplate: Value does not fall within the expected range

### ISSUE THREAD

**Issue by JPKotis** [SharePoint/PnP-Sites-Core#2417]
Thank you for reporting an issue or suggesting an enhancement. We appreciate your feedback - to help the team to understand your needs, please complete the below template to ensure we have the necessary details to assist you. If you have a actual question, we would ask you to use [SharePoint Developer Group](http://aka.ms/sppnp-community) at Microsoft Tech Community. Thanks!

#### Which PnP repository should you use to report the issue? 
*Please navigate to the appropriate repository by clicking on the link below and create your issue there. You can delete this section after you have navigated to the correct repository.*

- PnP Sites Core - https://github.com/OfficeDev/PnP-Sites-Core
- PnP PowerShell - https://github.com/OfficeDev/PnP-PowerShell

#### Category
[x ] Bug
[ ] Enhancement

#### Environment
[x ] Office 365 / SharePoint Online
[ ] SharePoint 2016
[ ] SharePoint 2013

If SharePoint on-premises, what's exact CU version: 

#### Expected or Desired Behavior
calling spcontext.Web.ApplyProvisioningTemplate succesfully

#### Observed Behavior
Time to tiime, not every time, calling spcontext.Web.ApplyProvisioningTemplate will fail during the Site Pages library (Sivuston sivut) provisioning phase with following error: 
Value does not fall within the expected range

The site template remains the exact the same each time, we just use the same template to apply site settings on multiple newly created SP Online Modern site collections. 

We have no idea whats happening in SPO side why the template fails one day, and then the next day its working again. It could work also after waiting for few hours after the first failure. 

This is happening currently on production environment on client's O365 subscription, we need to resolve the issue asap. 

#### Steps to Reproduce
Attached is the template file (removed the security section and navigation sections) and the produced log while applying the template.
[templatetoinvestigate.xml.txt](https://github.com/SharePoint/PnP-Sites-Core/files/3661592/templatetoinvestigate.xml.txt)
[loggertoinvestigate.txt](https://github.com/SharePoint/PnP-Sites-Core/files/3661595/loggertoinvestigate.txt)

**Comment by Pepozzo** [comment]
Same error here. In my case I've noticed that command is trying to restore data from a wrong list (that is also in another site).

**Comment by JPKotis** [comment]
It seems like it's caused by the fact that sometimes creating Groups site collection via GRAPH API is very slow and the site provisioning takes many minutes and until the site collection provisioning is completed fully (we can see the site in SP admin portal), we can't programmatically start applying templates to the site. 

I've tried to search a site "property" to give us a hint that the provisioning is complete, however haven't found anything yet. 

Using CSOM and SpoOperation there's "IsComplete" property that most likely works with CSOM, but cannot be used when creating Groups Site via Graph API (or is there a workaround for this?)

So at this point the only way how to make this work is to put the provisioning code "on hold" for several minutes (+ 5 minutes) after GRAPH API call and create a try-catch loop and do some retrying in case of an error occurs while applying the site template.

**Comment by PaoloPia** [comment]
Hi, I can confirm that the problem is that fully provisioning a site takes a while (minutes). In the near future we will be able to better monitor when provisioning is completed. So far, I would suggest you to wait at least 5 mins before "touching" a new site.

**Comment by baskb** [comment]
I think this issue is related to a similar issue in PnP PowerShell:
https://github.com/SharePoint/PnP-PowerShell/issues/1180

For me applying the template before Taxonomy fields where available was causing the issue. I solved this by first adding the fields, then enabling the taxonomy feature, and only after that aply the rest of the template handlers.

**Comment by holylander** [comment]
@JPKotis I would recommend you to enable the **pnptracelog**, and try to apply the site template using the **pnp-powershell** module. Afterwards, inspect the log and find our which is the issue or template code line where your script fails. Sometimes it is because there is a missing field or site featue.

Here is a simple script to always start with a clean log, just add it before any other pnp-powershell line:

```powershell
Set-PnPTraceLog -off
$tracelogname="traceoutput-[sitename].txt"
$tracelogPath="./$tracelogname"
if (Test-Path -Path $tracelogPath ){
  Remove-Item -Path $tracelogPath -Force -ErrorAction Continue
}
Set-PnPTraceLog -On -LogFile $tracelogname -Level Debug -AutoFlush $true```
